### PR TITLE
Relax dimension ordering rules for dot_general.

### LINKED ISSUE #2972: Inefficient `moveaxis` in `dot_general` batching rule
When batching a `dot_general` primitive, the batch axis is [moved to the end](https://github.com/google/jax/blob/b543652332a98e5c15db04651417404513aa4d1a/jax/lax/lax.py#L2563) of the array. Right now, this can cause problems for high-rank tensors because the order of axes affects the way XLA compiles the operation (in particular, can result in padding out the moved batched axis to a multiple of 128, instead of using one of the other larger axes).

Instead of moving the batch axis to the end, it should be possible to keep it in its original location, which would give the user more control over how XLA compiles the computation.

A repro that produces this behavior:
```
x = jnp.zeros((16, 3, 256, 256))
y = jnp.zeros((2, 4, 256))
z = jnp.zeros((16, 3, 4, 256, 2048))

def go(x, y, z):
  return jnp.einsum("vnn,sn,vsnj->nj", x, y, z)

@jax.jit
def repro(x, y, z):
  go_bat = jax.vmap(jax.vmap(go, (0, None, 0)), (None, 0, None))
  return go_bat(x, y, z)

repro(x, y, z)
```

### PR DESCRIPTION
**Pull request by hawkinsp**
JAX currently requires that batch dimensions appear first and contiguously in the arguments to dot_general. However, XLA does not require this; relax JAX's checks so that it also allows batch dimensions in arbitrary positions.

Since batch dimensions are now allowed in arbitrary positions, it's not hard to
generalize the dot_general batching rule to avoid performing any transposes. Fixes #2972 .

In passing, also move the bool/int dot expansion into the XLA translation rule. The expansion inside the `lax.dot_general()` wrapper predated the existence of (or at least my knowledge of) `xla.lower_fun()`.

### CODE REVIEW

**Review by jekbradbury — APPROVED** [review]
Thanks! I'm sorry I dropped the ball on this.

Also, last year I added a test [here](https://github.com/google/jax/pull/1331/files#diff-60f1c5fe94df9c63c0ea8aec5a29584e) to assert that vmap didn't add any broadcasts in certain cases; if vmap can now always take one `dot_general` to one `dot_general` we could add an assertion to that effect in the tests in `lax_vmap_test` instead (and delete the other one).

**Comment by hawkinsp** [comment]
> Thanks! I'm sorry I dropped the ball on this.
> 
> Also, last year I added a test [here](https://github.com/google/jax/pull/1331/files#diff-60f1c5fe94df9c63c0ea8aec5a29584e) to assert that vmap didn't add any broadcasts in certain cases; if vmap can now always take one `dot_general` to one `dot_general` we could add an assertion to that effect in the tests in `lax_vmap_test` instead (and delete the other one).

The jaxpr has some call structure to it so instead I added an assertion that `transpose` doesn't appear in the jaxpr, rather than testing for a particular structure.
